source/httpfs/driver.go

```diff
@@ -0,0 +1,159 @@
+package httpfs
+
+import (
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"path"
+
+	"github.com/golang-migrate/migrate/v4/source"
+)
+
+// driver is a migration source driver for reading migrations from
+// http.FileSystem instances. It implements source.Driver interface and can be
+// used as a migration source for the main migrate library.
+type driver struct {
+	migrations *source.Migrations
+	fs         http.FileSystem
+	path       string
+}
+
+// New creates a new migrate source driver from a http.FileSystem instance and a
+// relative path to migration files within the virtual FS. It is identical to
+// the WithInstance() function except it will delay any errors on fist usage of
+// this driver. This reduces the number of error handling branches in clients of
+// this package without compromising on correctness.
+func New(fs http.FileSystem, path string) source.Driver {
+	d, err := newDriver(fs, path)
+	if err != nil {
+		return &failedDriver{err}
+	}
+	return d
+}
+
+// WithInstance creates new migrate source driver from a http.FileSystem
+// instance and a relative path to migration files within the virtual FS.
+func WithInstance(fs http.FileSystem, path string) (source.Driver, error) {
+	return newDriver(fs, path)
+}
+
+func newDriver(fs http.FileSystem, path string) (*driver, error) {
+	root, err := fs.Open(path)
+	if err != nil {
+		return nil, err
+	}
+
+	files, err := root.Readdir(0)
+	if err != nil {
+		_ = root.Close()
+		return nil, err
+	}
+	if err = root.Close(); err != nil {
+		return nil, err
+	}
+
+	ms := source.NewMigrations()
+	for _, file := range files {
+		if file.IsDir() {
+			continue
+		}
+
+		m, err := source.DefaultParse(file.Name())
+		if err != nil {
+			continue // ignore files that we can't parse
+		}
+
+		if !ms.Append(m) {
+			return nil, fmt.Errorf("duplicate migration file: %v", file.Name())
+		}
+	}
+
+	return &driver{
+		fs:         fs,
+		path:       path,
+		migrations: ms,
+	}, nil
+}
+
+// Open is part of source.Driver interface implementation. It always returns
+// error because http.FileSystem must be provided by the user of this package
+// and created using WithInstance() function.
+func (h *driver) Open(url string) (source.Driver, error) {
```

Comment:
It's cleaner to not have `httpfs` implement the `source.Driver` interface than to have a method that always returns an error. e.g. take advantage of the type system

You'll need to implement a dummy source driver for unit tests and also note in the docs and README that `httpfs` isn't actually a source driver

Reply:
If this package would not provide an implementation of `source.Driver` it will be impossible to use it with `migrate.NewWithSourceInstance(...)`.

Would you prefer to split this package into two go packages. Internal one, and dummy implementation that exposes existing functionality?

Reply:
It doesn't make sense to make the `httpfs.HTTPFS` struct internal since anyone should be able to develop a driver w/o upstreaming it.

We could provide a `http.FileSystem` pass-through implementation. Maybe call it `httpfs.HTTPFSPassThrough`?